pkg/fqdn/name_manager.go

```diff
@@ -207,11 +237,18 @@ func (n *NameManager) ForceGenerateDNS(ctx context.Context, namesToRegen []strin
 		log.WithField(logfields.DNSName, namesMissingIPs).
 			Debug("No IPs to insert when generating DNS name selected by ToFQDN rule")
 	}
-	n.Mutex.Unlock()
 
 	// emit the new rules
-	return n.config.
-		UpdateSelectors(ctx, selectorIPMapping, namesMissingIPs)
+	errCh = make(chan error, 1)
+	n.updateQueue <- func() {
+		defer close(errCh)
+		wg, err := n.config.UpdateSelectors(ctx, selectorIPMapping, namesMissingIPs)
+		if err != nil {
+			errCh <- err
+		}
+		wg.Wait()
+	}
+	return errCh
```

Comment:
nit: If you still need to touch the code otherwise, consider adding a small function to do this to avoid the code duplication between this and `UpdateGenerateDNS()`.